### PR TITLE
fixed: delete attribute even if related collection does not exist

### PR DESCRIPTION
**Pull request by Akshay-Rana-Gujjar**
## What does this PR do?

Delete attribute even if related collection does not exist.

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Screenshots may also be helpful.)

## Related PRs and Issues

- https://github.com/appwrite/appwrite/issues/6012

## Checklist

- [x] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [ ] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

**Review by fanatic75 — COMMENTED** [review]
@Akshay-Rana-Gujjar , Let's add a test case for this scenario

**Comment by Akshay-Rana-Gujjar** [comment]
Hey @fanatic75, I have added the test case, please review 🙏

**Comment by Akshay-Rana-Gujjar** [comment]
Hey @fanatic75 , any update on this?

**Comment by Akshay-Rana-Gujjar** [comment]
@fanatic75 please review 🙏

**Comment by Akshay-Rana-Gujjar** [comment]
Hey @stnguyen90, Could you please help to review this PR.

File: tests/e2e/Services/Databases/DatabasesBase.php [hunk 1]
```diff
@@ -4216,4 +4216,85 @@ public function testUpdateWithExistingRelationships(array $data): void
 
         $this->assertEquals(200, $update['headers']['status-code']);
     }
+
+    /**
+     * @depends testCreateDatabase
+     * @param array $data
+     * @return void
+     * @throws \Exception
```
Comment (fanatic75) [inline tests/e2e/Services/Databases/DatabasesBase.php:4224]: Test case is not throwing exception anywhere, let's remove this

File: tests/e2e/Services/Databases/DatabasesBase.php [hunk 2]
```diff
@@ -4216,4 +4216,85 @@ public function testUpdateWithExistingRelationships(array $data): void
...
+        $level1 = $this->client->call(Client::METHOD_POST, '/databases/' . $databaseId . '/collections', array_merge([
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+            'x-appwrite-key' => $this->getProject()['apiKey']
+        ]), [
+            'collectionId' => ID::unique(),
+            'name' => 'Level 1',
+            'documentSecurity' => true,
+            'permissions' => [
+                Permission::create(Role::user($this->getUser()['$id'])),
```
Comment (fanatic75) [inline tests/e2e/Services/Databases/DatabasesBase.php:4240]: let's remove line break here

File: tests/e2e/Services/Databases/DatabasesBase.php [hunk 3]
```diff
@@ -4216,4 +4216,85 @@ public function testUpdateWithExistingRelationships(array $data): void
...
+        $level2 = $this->client->call(Client::METHOD_POST, '/databases/' . $databaseId . '/collections', array_merge([
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+            'x-appwrite-key' => $this->getProject()['apiKey']
+        ]), [
+            'collectionId' => ID::unique(),
+            'name' => 'Level 2',
+            'documentSecurity' => true,
+            'permissions' => [
+                Permission::create(Role::user($this->getUser()['$id'])),
```
Comment (fanatic75) [inline tests/e2e/Services/Databases/DatabasesBase.php:4254]: let's fix indentation here too

Comment (Akshay-Rana-Gujjar) [reply tests/e2e/Services/Databases/DatabasesBase.php:4251]: indentation seems correct, I checked on my system

File: tests/e2e/Services/Databases/DatabasesBase.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (Akshay-Rana-Gujjar) [reply tests/e2e/Services/Databases/DatabasesBase.php]: Done

File: tests/e2e/Services/Databases/DatabasesBase.php [hunk 4]
```diff
@@ -4216,4 +4216,82 @@ public function testUpdateWithExistingRelationships(array $data): void
...
     }
+
+    /**
+     * @depends testCreateDatabase
+     */
+    public function testDeleteRelationAttributeIfRelatedCollectionDeleted(array $data): void
+    {
+        $databaseId = $data['databaseId'];
+
+        // Create level 1 collection
```
Comment (fanatic75) [inline tests/e2e/Services/Databases/DatabasesBase.php:4227]: let's remove this comment

File: tests/e2e/Services/Databases/DatabasesBase.php [hunk 5]
```diff
@@ -4216,4 +4216,82 @@ public function testUpdateWithExistingRelationships(array $data): void
...
+        ]), [
+            'collectionId' => ID::unique(),
+            'name' => 'Level 1',
+            'documentSecurity' => true,
+            'permissions' => [
+                Permission::create(Role::user($this->getUser()['$id'])),
+                Permission::read(Role::user($this->getUser()['$id'])),
+            ],
+        ]);
+        // Create level 2 collection
```
Comment (fanatic75) [inline tests/e2e/Services/Databases/DatabasesBase.php:4241]: let's add a line break here and remove this comment 
```suggestion

```

File: tests/e2e/Services/Databases/DatabasesBase.php [hunk 6]
```diff
@@ -4216,4 +4216,82 @@ public function testUpdateWithExistingRelationships(array $data): void
...
+        $this->assertEquals(false, $response['body']['array']);
+        $this->assertEquals('level2', $response['body']['key']);
+        $this->assertEquals('cascade', $response['body']['onDelete']);
+        $this->assertEquals($level2['body']['$id'], $response['body']['relatedCollection']);
+        $this->assertEquals('manyToOne', $response['body']['relationType']);
+        $this->assertEquals(false, $response['body']['required']);
+        $this->assertEquals(false, $response['body']['twoWay']);
+        $this->assertEquals('relationship', $response['body']['type']);
+
+        // Delete level 2 collection
```
Comment (fanatic75) [inline tests/e2e/Services/Databases/DatabasesBase.php:4279]: let's only keep line break and no comment, seems straight forward

**Review by fanatic75 — COMMENTED** [review]
Just fix these small changes
Otherwise looks good to me. Good work!

File: tests/e2e/Services/Databases/DatabasesBase.php (hunk 6 above)
Comment (Akshay-Rana-Gujjar) [reply tests/e2e/Services/Databases/DatabasesBase.php]: Done

File: tests/e2e/Services/Databases/DatabasesBase.php (hunk 5 above)
Comment (Akshay-Rana-Gujjar) [reply tests/e2e/Services/Databases/DatabasesBase.php]: Done

File: tests/e2e/Services/Databases/DatabasesBase.php (hunk 4 above)
Comment (Akshay-Rana-Gujjar) [reply tests/e2e/Services/Databases/DatabasesBase.php]: Done

**Review by fanatic75 — APPROVED** [review]
LGTM, @abnegate  please review

**Comment by Akshay-Rana-Gujjar** [comment]
@stnguyen90 Can you please help with this merge it is almost 1 month and It is open 😢

**Comment by Akshay-Rana-Gujjar** [comment]
Hey @Haimantika & @tessamero, could you please help merge this MR? It has been open for quite some time.

**Review by abnegate — COMMENTED** [review]
How do you get to the state where the related collection doesn't exist but the relationship does? Are there steps to reproduce?

This sounds like the bug might be related to deleting a collection itself (which should delete all of it's relationships)

**Comment by Akshay-Rana-Gujjar** [comment]
> How do you get to the state where the related collection doesn't exist but the relationship does? Are there steps to reproduce?
> 
> This sounds like the bug might be related to deleting a collection itself (which should delete all of it's relationships)

Yes, I followed the steps mentioned in [this issue](https://github.com/appwrite/appwrite/issues/6012).

👟 Reproduction steps

1. Create collection level1
2. Create collection level2
3. Create attribute in level1
    - type: relationship
    - one/two way: one way
    - key: level2
    - related collection: level2
    - relation: Many to One
    - on deleting: Cascade
4. Delete collection level2
5. Delete attribute level2 from collection level1

**Comment by fogelito** [comment]
I agree, we should not get to this point we have a hanging attribute, but at least we should be able to fix it if it happens

**Comment by Akshay-Rana-Gujjar** [comment]
> I agree, we should not get to this point we have a hanging attribute, but at least we should be able to fix it if it happens

Hey @fogelito, could you please explain what I need to do further to get approvals and merge this MR? Thanks.

**Comment by fogelito** [comment]
Hey, thanks a lot 🙏 , @abnegate can you give your review here, please?

**Comment by abnegate** [comment]
@Akshay-Rana-Gujjar The root cause of the issue is that when a collection is deleted, the relationships are not deleted from the `attributes` table for the related collection, but are deleted from the metadata. So when trying to delete it manually later, it can't be found (in metadata) and gets `stuck`. Can we pivot the fix to address that instead?

**Comment by stnguyen90** [comment]
Closing as we've updated the logic to delete the related attributes when a collection is deleted.

**Pull request closed without merging**